pkg/pool-hooks/contracts/StableSurgeHook.sol

```diff
@@ -187,7 +191,100 @@ contract StableSurgeHook is BaseHooks, VaultGuard, SingletonAuthentication {
+        return (true, getSurgeSwapFeePercentage(params, pool, staticSwapFeePercentage));
+    }
+
+    function getSurgeSwapFeePercentage(PoolSwapParams calldata params, address pool, uint256 staticSwapFeePercentage) public view returns (uint256) {
+        uint256 numTokens = params.balancesScaled18.length;
+
+        uint256 amountCalculatedScaled18 = StablePool(pool).onSwap(params);
+
+        uint256[] memory newBalances = new uint256[](numTokens);
+        ScalingHelpers.copyToArray(params.balancesScaled18, newBalances);
+
+        if (params.kind == SwapKind.EXACT_IN) {
+            newBalances[params.indexIn] += params.amountGivenScaled18;
+            newBalances[params.indexOut] -= amountCalculatedScaled18;
+        } else {
+            newBalances[params.indexIn] += amountCalculatedScaled18;
+            newBalances[params.indexOut] -= params.amountGivenScaled18;
+        }
+
+        return _getSurgeFeePercentage(params, pool, staticSwapFeePercentage, newBalances);
+    }
+
+    /// @inheritdoc IHooks
+    function onAfterAddLiquidity(
+        address,
+        address pool,
+        AddLiquidityKind kind,
+        uint256[] memory amountsInScaled18,
+        uint256[] memory amountsInRaw,
+        uint256,
+        uint256[] memory balancesScaled18,
+        bytes memory
+    ) public view override returns (bool success, uint256[] memory hookAdjustedAmountsInRaw) {
+        hookAdjustedAmountsInRaw = amountsInRaw; // don't adjust amounts in.
+
+        // Proportional add is always fine.
+        if (kind == AddLiquidityKind.PROPORTIONAL) {
+            return (true, hookAdjustedAmountsInRaw);
+        }
+
+        // Rebuild old balances before adding liquidity.
+        uint256[] memory oldBalancesScaled18 = new uint256[](balancesScaled18.length);
+        for (uint256 i = 0; i < balancesScaled18.length; ++i) {
+            oldBalancesScaled18[i] = balancesScaled18[i] - amountsInScaled18[i];
+        }
+
+        SurgeFeeData memory surgeFeeData = _surgeFeePoolData[pool];
+        uint256 newTotalImbalance = StableSurgeMedianMath.calculateImbalance(balancesScaled18);
+
+        bool isSurging = _isSurging(
+            surgeFeeData,
+            oldBalancesScaled18,
+            newTotalImbalance
+        );
+
+        // If we're not surging, it's fine to proceed; otherwise halt execution by returning false.
+        return (isSurging == false, hookAdjustedAmountsInRaw);
+    }
+
+    /// @inheritdoc IHooks
+    function onAfterRemoveLiquidity(
+        address,
+        address pool,
+        RemoveLiquidityKind kind,
+        uint256,
+        uint256[] memory amountsOutScaled18,
+        uint256[] memory amountsOutRaw,
+        uint256[] memory balancesScaled18,
+        bytes memory
+    ) public view override returns (bool success, uint256[] memory hookAdjustedAmountsOutRaw) {
+        hookAdjustedAmountsOutRaw = amountsOutRaw; // don't adjust amounts out.
```

Comment:
Same as above; could remove `hookAdjustedAmountsOutRaw` and just return `amountsOutRaw`